module/input/SensorFilter/src/SensorFilter.cpp

```diff
@@ -57,174 +37,337 @@ namespace module::input {
+    void SensorFilter::update_kinematics(std::unique_ptr<Sensors>& sensors, const RawSensors& raw_sensors) {
+        // Convert the sensor joint angles to a configuration vector
+        Eigen::Matrix<double, n_servos, 1> q = sensors_to_configuration<double, n_servos>(sensors);
+
+        // **************** Kinematics ****************
+        // Htx is a map from FrameID to homogeneous transforms from each frame to the torso
+        std::vector<Eigen::Isometry3d> fk = tinyrobotics::forward_kinematics(nugus_model, q);
+        auto Htx                          = forward_kinematics_to_servo_map(fk);
+        for (const auto& entry : Htx) {
+            sensors->Htx[entry.first] = entry.second.matrix();
+        }
+
+        // **************** Centre of Mass  ****************
+        sensors->rMTt = tinyrobotics::center_of_mass(nugus_model, q);
+
+        // **************** Foot Down Information ****************
+        sensors->feet[BodySide::RIGHT].down = true;
+        sensors->feet[BodySide::LEFT].down  = true;
+        const Eigen::Isometry3d Htr(sensors->Htx[FrameID::R_ANKLE_ROLL]);
+        const Eigen::Isometry3d Htl(sensors->Htx[FrameID::L_ANKLE_ROLL]);
+        const Eigen::Isometry3d Hlr = Htl.inverse() * Htr;
+        const Eigen::Vector3d rRLl  = Hlr.translation();
+        switch (cfg.foot_down.method()) {
+            case FootDownMethod::Z_HEIGHT:
+                if (rRLl.z() < -cfg.foot_down.threshold()) {
+                    sensors->feet[BodySide::LEFT].down = false;
+                }
+                else if (rRLl.z() > cfg.foot_down.threshold()) {
+                    sensors->feet[BodySide::RIGHT].down = false;
+                }
+                break;
+            case FootDownMethod::FSR:
+                // Determine if any two diagonally opposite FSRs are in contact with the ground, which is either fsr1
+                // and fsr3, or fsr2 and fsr4. A FSR is in contact with the ground if its value is greater than the
+                // certainty threshold
+                sensors->feet[BodySide::LEFT].down = (raw_sensors.fsr.left.fsr1 > cfg.foot_down.threshold()
+                                                      && raw_sensors.fsr.left.fsr3 > cfg.foot_down.threshold())
+                                                     || (raw_sensors.fsr.left.fsr2 > cfg.foot_down.threshold()
+                                                         && raw_sensors.fsr.left.fsr4 > cfg.foot_down.threshold());
+                sensors->feet[BodySide::RIGHT].down = (raw_sensors.fsr.right.fsr1 > cfg.foot_down.threshold()
+                                                       && raw_sensors.fsr.right.fsr3 > cfg.foot_down.threshold())
+                                                      || (raw_sensors.fsr.right.fsr2 > cfg.foot_down.threshold()
+                                                          && raw_sensors.fsr.right.fsr4 > cfg.foot_down.threshold());
```

Comment:
I felt this was a bit long and hard to read so here's an alternative that copilot suggested to me

```suggestion
                bool isFootDown(const Fsr& fsr, const Threshold& threshold) {
                    return (fsr.fsr1 > threshold && fsr.fsr3 > threshold)
                           || (fsr.fsr2 > threshold && fsr.fsr4 > threshold);
                }

                sensors->feet[BodySide::LEFT].down  = isFootDown(raw_sensors.fsr.left, cfg.foot_down.threshold());
                sensors->feet[BodySide::RIGHT].down = isFootDown(raw_sensors.fsr.right, cfg.foot_down.threshold());
```